### ISSUE TITLE
Error on valid FlatMap

### ISSUE THREAD

**Issue by Kixiron** [vmware/differential-datalog#938]
With this code

```prolog
typedef Row = Row { n: usize }
typedef Col = Col { n: usize }

typedef Cell = Cell { row: Row, col: Col }

typedef Value = Value { n: usize }

input relation InitialCells(cell: Cell, value: Value)

relation SolvedCells(cell: Cell)
SolvedCells(cell) :- InitialCells(cell, _).

relation PossibleValues(cell: Cell, value: Value)
PossibleValues(cell, value) :- (var cell, var value) = FlatMap(all_possibilities()).

function all_possibilities(): Vec<(Cell, Value)> {
    var possibilities = vec_with_capacity(9 * 9 * 9);
    for (row in range_vec(1, 9, 1)) {
        for (col in range_vec(1, 9, 1)) {
            for (value in range_vec(1, 9, 1)) {
                var cell = Cell {
                    .row = Row { row },
                    .col = Col { col },
                };
                var value = Value { value };

                possibilities.push((cell, value));
            }
        }
    };

    possibilities
}

```

DDlog gives this error

```
> ddlog -i code.dl

error: code.dl:14.1-16.1: Rule must start with a positive literal
PossibleValues(cell, value) :- (var cell, var value) = FlatMap(all_possibilities()).
^
```

It seems the error comes from [`ruleValidate`](https://github.com/vmware/differential-datalog/blob/master/src/Language/DifferentialDatalog/Validate.hs#L304-L313) but I'm not sure why since this was explicitly implemented in a154c94d49dd695270deba01026e6a8564013022

**Comment by ryzhyk** [comment]
This is indeed invalid DDlog at the moment.  This is the same issue as #770.

**Comment by Kixiron** [comment]
How is the flat map a negative rule, the error makes sense for a negation but not really on this

**Comment by ryzhyk** [comment]
This applies to any rule that doesn't start with a positive literal.  Injecting a singleton relation in the body of such a rule makes it possible to evaluate it using the existing compilation algorithm.  In particular, we compile `FlatMap`, down to DD's `flatmap` operator, which requires an input collection.  `Singleton` would serve as such an input collection. Your rule would become:

```
PossibleValues(cell, value) :- Singleton(),
                                              (var cell, var value) = FlatMap(all_possibilities()).
```

**Comment by Kixiron** [comment]
Ah, gotcha. As a side note we should probably special case this kind of situation into a `.to_stream()`/`.as_collection()`

**Comment by mbudiu-vmw** [comment]
Stdlib should include a unit collection with am empty tuple that the compiler can always use for a positive atom

________________________________

**Comment by ryzhyk** [comment]
> Ah, gotcha. As a side note we should probably special case this kind of situation into a `.to_stream()`/`.as_collection()`

Right, this particular situation and possibly some others can be handled that way. But I think `Singleton` will work well too. We will declare it as

```
relation Singleton()
Singleton().
```

This will populate it with a single record and close the input handle, so there should be little or no overhead to evaluating rules like yours.

**Comment by ryzhyk** [comment]
Closed via #953
